NuRadioReco/modules/channelCWNotchFilter.py

```diff
@@ -0,0 +1,227 @@
+on peaks in frequency spectrum
+"""
+
+
+def find_frequency_peaks_from_trace(trace : np.ndarray, fs : float, threshold : float = 4):
+    """
+    Function fo find the frequency peaks in the real fourier transform of the input trace.
+
+    Parameters
+    ----------
+    trace : np.ndarray
+        Waveform 
+    fs : float
+        Sampling frequency, (input should be taking from the channel object)
+    threshold : float, default = 4
+        Threshold for peak definition. A peak is defined as a point in the frequency spectrum
+        that exceeds threshold * rms(real fourier transform)
+    
+    Returns
+    -------
+    freq_peaks : np.ndarray
+        Frequencies at which a peak was found
+    """
+    freq = np.fft.rfftfreq(len(trace), d=1/fs)
+    ft = fft.time2freq(trace, fs)
+    
+    freq_peaks = find_frequency_peaks(freq, ft, fs=fs, threshold=threshold)
+
+    return freq_peaks
+
+def find_frequency_peaks(freq: np.ndarray, spectrum : np.ndarray, threshold : float = 4):
+    """
+    Function fo find the frequency peaks in the real fourier transform of the input trace.
+
+    Parameters
+    ----------
+    freq : np.ndarray
+        Frequencies of a NuRadio time trace
+    spectrum : np.ndarray
+        Spectrum of a NuRadio time trace
+    threshold : float, default = 4
+        Threshold for peak definition. A peak is defined as a point in the frequency spectrum
+        that exceeds threshold * rms(real fourier transform)
+    
+    Returns
+    -------
+    freq : np.ndarray
+        Frequencies at which a peak was found
+    """
+    
+    rms = np.sqrt(np.mean(np.abs(spectrum)**2))
+    peak_idxs = np.where(np.abs(spectrum) > threshold * rms)[0]
+
+    return freq[peak_idxs]
+
+
+def filter_cws(trace : np.ndarray, freq : np.ndarray, spectrum : np.ndarray, fs=3.2e9 * units.Hz, quality_factor=1e3, threshold=4):
+    """
+    Function that applies a notch filter at the frequency peaks of a given time trace
+    using the scipy library
+
+    Parameters
+    ----------
+    trace : np.ndarray
+        waveform (shape: [2048])
+    freq : np.ndarray
+        Frequency of the trace's real fourier transform
+    spectrum:
+        the trace's real fourier transform
+    fs : float, default = 3.2e9 Hz
+        sampling frequency of the RNO-G DAQ
+    quality_factor : int, default = 1000
+        quality factor of the notch filter, defined as the ratio f0/bw, where f0 is the centre frequency
+        and bw the bandwidth of the filter at (f0,-3 dB)
+    threshold : int, default = 4
+        threshold for peak definition. A peak is defined as a point in the frequency spectrum
+        that exceeds threshold * rms(real fourier transform)
+
+    """
+    freqs = find_frequency_peaks(freq, spectrum, threshold=threshold)
+
+    if len(freqs):
+        notch_filters = [signal.iirnotch(freq, quality_factor, fs = fs) for freq in freqs]
+        trace_notched = signal.filtfilt(notch_filters[0][0], notch_filters[0][1], trace)
```

Comment:
If it does pad, I see two issues: 
1) you get a longer trace than you put in, which is perhaps surprising
2) you can get weird transients at the former edge of the waveform, where you'll essentially see the step response of the filter

Reply:
Hi Cosmin, thank you for the input. I am very much not experienced with signal analysis / filters so I might be asking some obvious questions in what follows.
Filtfit indeed pads the waveform. I looked in the documentation and it applies an "odd padding" (extending the waveform by copying a range of samples at the edge, rotated by 180 degrees). The trace that is returned is the same shape as the input trace so after filter application it cuts the padding. Is it a problem to use padding? In my naive estimation I thought the filter would lead to some artifacts at the edges when not using any padding? Do you mean there would be transients due to the step reponse if the the wavefrom was padded with 0's?

Reply:
If it does indeed crop the padding applied then that's good and we don't have to worry about that part. 

And yes, I'm somewhat worried about the step response of the filter at the edge after padding, since you'll be going from 0 to sometimes a non-zero value that sometimes will be large (i.e. > 3 sigma 0.3% of the time), but I guess as long as we never consider anything on the edge of the waveforms it's probably not a problem?

Reply:
For now I have set the padding to None (not sure how Scipy then arrives at the same trace length though), Odd padding is also an option as it basically "extends" the waveform. I do not have a feeling about which option is better.